odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+        """
+        if other is self:
+            return True
+        elif other not in self.space:
+            return False
+        else:
+            return bool((self.data == other.data).all())
+
+    def copy(self):
+        """Create an identical (deep) copy of this tensor.
+
+        Returns
+        -------
+        copy : `pygpu._array.ndgpuarray`
+            A deep copy.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> y = x.copy()
+        >>> y
+        rn(3, impl='cupy').element([ 1.,  2.,  3.])
+        >>> x == y
+        True
+        >>> x is y
+        False
+        """
+        return self.space.element(self.data.copy())
+
+    def __getitem__(self, indices):
+        """Access values of this tensor.
+
+        Parameters
+        ----------
+        indices : index expression
+            The position(s) that should be accessed.
+
+        Returns
+        -------
+        values : scalar or `pygpu._array.ndgpuarray`
+            The value(s) at the index (indices).
+
+        Examples
+        --------
+        Indexing rules follow roughly the Numpy style, as far (or "fancy")
+        as supported:
+
+        >>> r5 = odl.rn(5, impl='cupy')
+        >>> x = r5.element([1, 2, 3, 4, 5])
+        >>> x[1:4]
+        rn(3, impl='cupy').element([ 2.,  3.,  4.])
+        >>> x[::2]
+        rn(3, impl='cupy').element([ 1.,  3.,  5.])
+
+        The returned views are writable, so modificatons alter the
+        original array:
+
+        >>> view = x[1:4]
+        >>> view[:] = -1
+        >>> view
+        rn(3, impl='cupy').element([-1., -1., -1.])
+        >>> x
+        rn(5, impl='cupy').element([ 1., -1., -1., -1.,  5.])
+
+        Multi-indexing is also directly supported:
+
+        >>> tensors = odl.rn((2, 3), impl='cupy')
+        >>> x = tensors.element([[1, 2, 3],
+        ...                      [4, 5, 6]])
+        >>> x[1, 2]
+        6.0
+        >>> x[1]  # row with index 1
+        rn(3, impl='cupy').element([ 4.,  5.,  6.])
+        >>> view = x[:, ::2]
+        >>> view
+        rn((2, 2), impl='cupy').element(
+            [[ 1.,  3.],
+             [ 4.,  6.]]
+        )
+        >>> view[:] = [[0, 0],
+        ...            [0, 0]]
+        >>> x
+        rn((2, 3), impl='cupy').element(
+            [[ 0.,  2.,  0.],
+             [ 0.,  5.,  0.]]
+        )
+        """
+        arr = self.data[indices]
+        if arr.shape == ():
+            if arr.dtype.kind == 'f':
+                return float(np.asarray(arr))
+            elif arr.dtype.kind == 'c':
+                return complex(np.asarray(arr))
+            elif arr.dtype.kind in ('u', 'i'):
+                return int(np.asarray(arr))
+            else:
+                raise RuntimeError("no conversion for dtype {}"
+                                   "".format(arr.dtype))
+        else:
+            space = type(self.space)(arr.shape, dtype=self.dtype,
```

Comment:
weighting?

Reply:
I'll port the current implementation in `NumpyTensorSpace`, the proper solution will come with #1238 
Just in general: if you squash axes, there's no obvious way to propagate weightings.

Reply:
We need to settle what squashed axes mean. Imo it should set that dimension to "1" so to say, i.e. we keep `space[0:1].weighting` and `space[0].weighting` different.

Reply:
Well, as of now, i.e., before #909 is done (part of #1238) we can't know how to handle weighting constants when removing axes. If all axes stay intact, we can keep the constant. If an axis is removed, we have to fall back to default. Not so great, but yeah.
For array weighting, i.e., the weights have the same shape as the space, we can just index them the same way, so that's a bit easier.